### PR TITLE
feat: Provide Azure Data Explorer Scaler (#1488)

### PR DESCRIPTION
**Pull request by yasiboni-max**
Signed-off-by: Yarden Siboni <yardensib@gmail.com>

This pull request adds an Azure Data Explorer scaler. This is related to this issue: #1488 

Here is the keda-docs PR: https://github.com/kedacore/keda-docs/pull/658

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [x] Tests have been added
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Relates to #1488

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Thanks a ton, great addition!

File: tests/scalers/azure-data-explorer.test.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
+import * as fs from 'fs'
+import * as sh from 'shelljs'
+import * as tmp from 'tmp'
+import test from 'ava'
+
+const dataExplorerDb = process.env['TEST_DATA_EXPLORER_DB']
+const dataExplorerEndpoint = process.env['TEST_DATA_EXPLORER_ENDPOINT']
```
Comment (zroubalik) [inline tests/scalers/azure-data-explorer.test.ts:7]: Could you please add `AZURE_` prefix to these envs?

File: pkg/scalers/azure_data_explorer_scaler.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+		threshold, err := strconv.ParseInt(val, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("error parsing metadata. Details: can't parse threshold. Inner Error: %v", err)
+		}
+		metadata.Threshold = threshold
+	}
+
+	// Resolve metricName.
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
+		metadata.MetricName = kedautil.NormalizeString(fmt.Sprintf("%s-%s", adxName, val))
```
Comment (zroubalik) [inline pkg/scalers/azure_data_explorer_scaler.go:105]: We prefer to not expose `metricName` to the scaler config, some older scalers is doing that, but any new one shouldn't do that. It is internal implementation detail and users shouldn't be concerned. Please remove this.

File: pkg/scalers/azure/azure_data_explorer.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+	return client, nil
+}
+
+func getDataExplorerAuthConfig(metadata *DataExplorerMetadata) (*auth.AuthorizerConfig, error) {
+	var authConfig auth.AuthorizerConfig
+
+	if metadata.PodIdentity != "" {
+		authConfig = auth.NewMSIConfig()
+		azureDataExplorerLogger.Info("Creating Azure Data Explorer Client using Pod Identity")
```
Comment (zroubalik) [inline pkg/scalers/azure/azure_data_explorer.go:65]: We might want to print these messages on DEBUG level, to not flood log with messages.
Could you please convert all `.Info()` messages in this file to `.V(1).Info()`?

File: pkg/scalers/azure_data_explorer_scaler.go [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+	metadata := azure.DataExplorerMetadata{}
+
+	switch config.PodIdentity {
+	case kedav1alpha1.PodIdentityProviderAzure:
+		metadata.PodIdentity = string(config.PodIdentity)
+	case "", kedav1alpha1.PodIdentityProviderNone:
+		// Check whether pod identity is provided through metadata instead of TriggerAuthentication object
+		podIdentity, err := getParameterFromConfig(config, "podIdentity", true)
+		if err != nil {
+			dataExplorerLogger.Info("pod identity is not provided. trying to resolve clientId, clientSecret and tenantId.")
```
Comment (zroubalik) [inline pkg/scalers/azure_data_explorer_scaler.go:130]: Do you think this is wort printing in INFO level? Maybe DEBUG would be better, like the other messages? WDYT?

File: tests/scalers/azure-data-explorer.test.ts (hunk 1 above)
Comment (zroubalik) [reply tests/scalers/azure-data-explorer.test.ts:7]: `AZURE_DATA_EXPLORER_TEST_DB`
`AZURE_DATA_EXPLORER_TEST_ENDPOINT` 

maybe to something like this?

Comment (tomkerkhove) [reply tests/scalers/azure-data-explorer.test.ts:7]: That's what  I'm adding in https://github.com/kedacore/keda/pull/2643 indeed

Comment (zroubalik) [reply tests/scalers/azure-data-explorer.test.ts:7]: So these envs were added as:
```
       AZURE_DATA_EXPLORER_DB: ${{ secrets.AZURE_DATA_EXPLORER_DB }}
       AZURE_DATA_EXPLORER_ENDPOINT: ${{ secrets.AZURE_DATA_EXPLORER_ENDPOINT }}
```

Comment (yasiboni-max) [reply tests/scalers/azure-data-explorer.test.ts]: done

File: pkg/scalers/azure_data_explorer_scaler.go (hunk 2 above)
Comment (yasiboni-max) [reply pkg/scalers/azure_data_explorer_scaler.go]: done

File: pkg/scalers/azure/azure_data_explorer.go (hunk 3 above)
Comment (yasiboni-max) [reply pkg/scalers/azure/azure_data_explorer.go]: done

File: pkg/scalers/azure_data_explorer_scaler.go (hunk 4 above)
Comment (yasiboni-max) [reply pkg/scalers/azure_data_explorer_scaler.go]: done

File: tests/scalers/azure-data-explorer.test.ts [hunk 5]
```diff
@@ -0,0 +1,237 @@
+import * as fs from 'fs'
+import * as sh from 'shelljs'
+import * as tmp from 'tmp'
+import test from 'ava'
+
+const dataExplorerDb = process.env['AZURE_TEST_DATA_EXPLORER_DB']
+const dataExplorerEndpoint = process.env['AZURE_TEST_DATA_EXPLORER_ENDPOINT']
```
[the remark this answers is not in the document]
Comment (yasiboni-max) [reply tests/scalers/azure-data-explorer.test.ts]: Oh sorry, renamed them.

**Comment by yasiboni-max** [comment]
> /run-e2e azure-data-explorer* **Update:** You can check the progres [here](https://github.com/kedacore/keda/actions/runs/1854502319)

Failed with 403 Forbidden status code - `Principal 'aadapp=***;***' is not authorized to access database '***'`

This az cli [command](https://docs.microsoft.com/en-us/cli/azure/kusto/database?view=azure-cli-latest#az-kusto-database-add-principal) can help with that.

**Comment by zroubalik** [comment]
> > /run-e2e azure-data-explorer* **Update:** You can check the progres [here](https://github.com/kedacore/keda/actions/runs/1854502319)
> 
> Failed with 403 Forbidden status code - `Principal 'aadapp=***;***' is not authorized to access database '***'`
> 
> This az cli [command](https://docs.microsoft.com/en-us/cli/azure/kusto/database?view=azure-cli-latest#az-kusto-database-add-principal) can help with that.

@tomkerkhove PTAL

**Comment by yasiboni-max** [comment]
> > > /run-e2e azure-data-explorer* **Update:** You can check the progres [here](https://github.com/kedacore/keda/actions/runs/1854502319)
> 
> > 
> 
> > Failed with 403 Forbidden status code - `Principal 'aadapp=***;***' is not authorized to access database '***'`
> 
> > 
> 
> > This az cli [command](https://docs.microsoft.com/en-us/cli/azure/kusto/database?view=azure-cli-latest#az-kusto-database-add-principal) can help with that.
> 
> 
> 
> @tomkerkhove PTAL

Any update with the role assignments?

**Comment by tomkerkhove** [comment]
Sorry, I had missed this :/

**Comment by tomkerkhove** [comment]
It failed again but not sure why, or I am looking over it.

**Comment by yasiboni-max** [comment]
> It failed again but not sure why, or I am looking over it.

It failed to create new namespace.
Can you please check that there's no existing ns with the name `test-azure-data-explorer`?

**Comment by yasiboni-max** [comment]
> > It failed again but not sure why, or I am looking over it.
> 
> It failed to create new namespace. Can you please check that there's no existing ns with the name `test-azure-data-explorer-ns`?

@tomkerkhove, I pushed a commit that will create namespace only if it doesn't exist already.
Can you please run the e2e again?

**Comment by zroubalik** [comment]
> > > It failed again but not sure why, or I am looking over it.
> > 
> > 
> > It failed to create new namespace. Can you please check that there's no existing ns with the name `test-azure-data-explorer-ns`?
> 
> @tomkerkhove, I pushed a commit that will create namespace only if it doesn't exist already. Can you please run the e2e again?

I think it would be better if it could delete (clean) the namespace if it exists and recreate then.

**Comment by yasiboni-max** [comment]
> 

[done](https://github.com/kedacore/keda/pull/2627/commits/5135e2c7839f79e07cc7a6943c61d503bbe378fc)

**Comment by tomkerkhove** [comment]
It's still creating it if it doesn't exist though - I like Zbynek his suggestion more though.

**Comment by yasiboni-max** [comment]
[This way](https://github.com/kedacore/keda/pull/2627/commits/03379883e51a294acd11b881beb6015e16ed6a29) it will either clean the ns if it exists or create new one if it doesn't. WDYT?

**Comment by tomkerkhove** [comment]
@zroubalik Anything we need as well?

**Comment by tomkerkhove** [comment]
@yasiboni-max Would you mind rebasing main please?

**Review by zroubalik — APPROVED** [review]
LGTM

File: CHANGELOG.md [hunk 6]
```diff
@@ -32,6 +32,7 @@
 
 - **General:** Automatically release container image for ARM ([#2263]https://github.com/kedacore/keda/issues/2263))
 - **General:** Automatically run end-to-end tests on ARM ([#2262]https://github.com/kedacore/keda/issues/2262))
+- **Azure Data Explorer Scaler:** Add Azure Data Explorer Scaler ([#2627](https://github.com/kedacore/keda/pull/2627))
```
Comment (tomkerkhove) [inline CHANGELOG.md:35]: Not - New scalers fall under general, see our contributions guide

**Comment by yasiboni-max** [comment]
Not sure why changing one line in CHANGELOG.md caused snyk's security check to fail, and it seems like it don't have enough permissions to see the full report. @tomkerkhove - can you help me with that please?

**Review by tomkerkhove — APPROVED** [review]
LGTM

**Comment by tomkerkhove** [comment]
> Not sure why changing one line in CHANGELOG.md caused snyk's security check to fail, and it seems like I don't have enough permissions to see the full report. @tomkerkhove Tom Kerkhove FTE - can you help me with that please?

It failed to process it, no worries. Thanks for checking (unfortunately we can't make reports public due to Snyk limitation)

**Pull request merged**